flink-connectors/flink-hbase/src/main/java/org/apache/flink/addons/hbase/HBaseTableSourceInputFormat.java

```diff
@@ -19,99 +19,113 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.io.IOException;
-import java.math.BigDecimal;
-import java.math.BigInteger;
 import java.util.ArrayList;
-import java.util.Date;
 import java.util.List;
+import java.util.Map;
 
 /**
  * {@link InputFormat} subclass that wraps the access for HTables. Returns the result as {@link Row}
  */
-public class HBaseTableSourceInputFormat extends RichInputFormat<Row, TableInputSplit> implements ResultTypeQueryable<Row> {
+public class HBaseTableSourceInputFormat extends TableInputFormat<Row> implements ResultTypeQueryable<Row> {
 
 	private static final long serialVersionUID = 1L;
 
 	private static final Logger LOG = LoggerFactory.getLogger(HBaseTableSourceInputFormat.class);
 	private String tableName;
-	private TypeInformation[] fieldTypeInfos;
-	private String[] fieldNames;
-	private transient Table table;
-	private transient Scan scan;
 	private transient Connection conn;
-	private ResultScanner resultScanner = null;
-
-	private byte[] lastRow;
-	private int scannedRows;
-	private boolean endReached = false;
-	private org.apache.hadoop.conf.Configuration conf;
-	private static final String COLON = ":";
+	private transient org.apache.hadoop.conf.Configuration conf;
+	private HBaseTableSchema schema;
 
-	public HBaseTableSourceInputFormat(org.apache.hadoop.conf.Configuration conf, String tableName, String[] fieldNames, TypeInformation[] fieldTypeInfos) {
-		this.conf = conf;
+	public HBaseTableSourceInputFormat(org.apache.hadoop.conf.Configuration conf, String tableName, HBaseTableSchema schema) {
 		this.tableName = tableName;
-		this.fieldNames = fieldNames;
-		this.fieldTypeInfos = fieldTypeInfos;
+		this.conf = conf;
+		this.schema = schema;
 	}
 
 	@Override
 	public void configure(Configuration parameters) {
 		LOG.info("Initializing HBaseConfiguration");
 		connectToTable();
 		if(table != null) {
-			scan = createScanner();
+			scan = getScanner();
 		}
 	}
 
-	private Scan createScanner() {
+	@Override
+	protected Scan getScanner() {
+		// TODO : Pass 'rowkey'. For this we need FilterableTableSource
 		Scan scan = new Scan();
-		for(String field : fieldNames) {
+		Map<String, List<Pair>> familyMap = schema.getFamilyMap();
+		for(String family : familyMap.keySet()) {
 			// select only the fields in the 'selectedFields'
-			String[] famCol = field.split(COLON);
-			scan.addColumn(Bytes.toBytes(famCol[0]), Bytes.toBytes(famCol[1]));
+			List<Pair> colDetails = familyMap.get(family);
+			for(Pair<String, TypeInformation<?>> pair : colDetails) {
+				scan.addColumn(Bytes.toBytes(family), Bytes.toBytes(pair.getFirst()));
+			}
 		}
 		return scan;
 	}
 
+	@Override
+	public String getTableName() {
+		return tableName;
+	}
+
+	@Override
+	protected Row mapResultToTuple(Result res) {
+		List<Object> values = new ArrayList<Object>();
+		int i = 0;
+		Map<String, List<Pair>> familyMap = schema.getFamilyMap();
+		Row[] rows = new Row[familyMap.size()];
```

Comment:
Better to declare `rows` as `Object[]` to avoid confusing whether `rows` is a varargs or non-varargs.